pallets/phala/src/fat.rs

```diff
@@ -0,0 +1,284 @@
+		CodeUploaded(CodeHash<T>),
+		PubkeyAvailable(ContractId, ContractPublicKey),
+		Instantiating(ContractId, ContractClusterId, T::AccountId),
+		Instantiated(ContractId, ContractClusterId, H256),
+		InstantiationFailed(ContractId, ContractClusterId, H256),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		CodeNotFound,
+		ContractClusterNotFound,
+		DuplicatedContract,
+		DuplicatedDeployment,
+		NoWorkerSpecified,
+		InvalidSender,
+		WorkerNotFound,
+	}
+
+	type CodeHash<T> = <T as frame_system::Config>::Hash;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		T: crate::mq::Config + crate::registry::Config,
+	{
+		#[pallet::weight(0)]
+		pub fn upload_code(origin: OriginFor<T>, code: Vec<u8>) -> DispatchResult {
+			ensure_signed(origin)?;
+			let code_hash = T::Hashing::hash(&code);
+			ContractCode::<T>::insert(&code_hash, &code);
+			Self::deposit_event(Event::CodeUploaded(code_hash));
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn upload_code_to_cluster(
+			origin: OriginFor<T>,
+			code: Vec<u8>,
+			cluster_id: ContractClusterId,
+		) -> DispatchResult {
+			let origin: T::AccountId = ensure_signed(origin)?;
+			// TODO.shelven: check permission?
+			Self::push_message(ContractOperation::UploadCodeToCluster {
+				origin,
+				code,
+				cluster_id,
+			});
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn instantiate_contract(
+			origin: OriginFor<T>,
+			// #[pallet::compact] endowment: BalanceOf<T>,
+			// #[pallet::compact] gas_limit: Weight,
+			code_index: CodeIndex<CodeHash<T>>,
+			data: Vec<u8>,
+			salt: Vec<u8>,
+			deploy_to: DeployTarget,
+		) -> DispatchResult {
+			let deployer = ensure_signed(origin)?;
+
+			match code_index {
+				CodeIndex::NativeCode(_) => {}
+				CodeIndex::WasmCode(code_hash) => {
+					ensure!(
+						ContractCode::<T>::contains_key(code_hash),
+						Error::<T>::CodeNotFound
+					);
+				}
+			}
+
+			let (cluster_id, deploy_workers) = match deploy_to {
+				DeployTarget::Cluster(cluster_id) => {
+					ensure!(
+						ClusterWorkers::<T>::contains_key(cluster_id),
+						Error::<T>::ContractClusterNotFound
+					);
+					let workers = ClusterWorkers::<T>::get(cluster_id).expect("checked; qed.");
+					(cluster_id, workers)
+				}
+				DeployTarget::NewGroup(deploy_workers) => {
+					ensure!(deploy_workers.len() > 0, Error::<T>::NoWorkerSpecified);
+
+					let counter = ContractClusterCounter::<T>::mutate(|counter| {
+						*counter += 1;
+						*counter
+					});
+					let cluster_id = ContractClusterId::from_low_u64_be(counter);
+					(cluster_id, deploy_workers)
+				}
+			};
+
+			let mut workers = Vec::new();
+			for worker in deploy_workers.clone().into_iter() {
+				let worker_info =
+					registry::Workers::<T>::try_get(&worker).or(Err(Error::<T>::WorkerNotFound))?;
+				workers.push(WorkerIdentity {
+					pubkey: worker_info.pubkey,
+					ecdh_pubkey: worker_info.ecdh_pubkey,
+				});
+			}
+			if !ClusterWorkers::<T>::contains_key(cluster_id) {
```

Comment:
This will only happen when deploying to a new cluster. Then we don't need to check `ClusterWorkers` storage map here. Instead, we can check `deploy_to` by `matches!()`. In this way we can avoid one additional read.

Reply:
I delay the invocation here to ensure each given worker is presented on-chain. Added a flag to reduce this additional read.